Pipeline and stages logic

vladsaveliev opened the pull request:
As part of the initial PR, adding the basic logic for building pipelines: `Pipeline` class, `Stage` class and `@stage` decorator, along with auxiliary classes.

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+
+        self.status_reporter = None
+        input_provider = None
+        if (input_provider_type == InputProviderType.SMDB 
+                or status_reporter_type == StatusReporterType.SMDB):
```
I think Black would format this differently, probably something like:

```
if (
    <condition 1>
    or <condition 2>
):
    pass

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+        else:
+            raise PipelineError(f'Unsupported storage policy {storage_policy}')
+
+        self.cohort = Cohort(
```
Why does a pipeline represent a Cohort, couldn't it be a Target.

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+        # Second round - actually adding jobs from the stages.
+        for i, (_, stage_) in enumerate(self._stages_dict.items()):
+            if not stage_.skipped:
+                logger.info(f'*' * 60)
```
```suggestion
                logger.info('*' * 60)
```

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+                    logger.info(f'Last stage is {stage_.name}, stopping here')
+                    break
+
+    def get_datasets(self, only_active: bool = True) -> list[Dataset]:
```
This feeling like it pollutes that Pipeline definition, but I don't have a good grasp.

illusional commented on cpg_pipes/pipeline/stages.py:
```diff
@@ -0,0 +1,121 @@
+"""
```
I think this file should be called `stage_targets.py`, I got quite confused about the naming while reviewing this PR.

vladsaveliev replied:
Renamed to `stage_subclasses.py`

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+_ALL_DEFINED_STAGES = []
+
+
+def stage(
```
The order of declarations in this File is a bit strange, to me it would be something like:

- StageInput / StageOutput
- Stage
- stage decorator
- Pipeline

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+    def __init__(
+        self,
+        data: StageOutputData,
+        stage: 'Stage',
```
Does the StageOutput / Input need a definition to the stage it's in? I can see it being relevant, but function definitions below do pass in a stage.

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+
+    def __init__(self, stage: 'Stage'):
+        self.stage = stage
+        self._results_by_target_by_stage: dict[str, dict[str, StageOutput]] = {}
```
This variable name could be clarified with by calling it `_result_by_stage_then_by_target`, which is what the usage implies.

illusional replied:
Just thinking more about this, why do the results have to be keyed by Stage?

vladsaveliev replied:
For the scenarios when a stage depends on multiple other stages

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+        fun: Callable,
+        stage: StageDecorator,
+    ):
+        if stage.__name__ not in [s.name for s in self.stage.required_stages]:
```
What are the differences between `self.stage` and `stage` here?

vladsaveliev replied:
_this_ stage vs the dependant stage.

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+        """
+        return self._each(fun=(lambda r: r.as_resource(id=id)), stage=stage)
+
+    def as_dict_by_target(
```
What is the behaviour of these methods if there are mixed output types?

illusional commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+    def as_path_by_target(
+        self,
+        stage: StageDecorator,
+        id: str | None = None,
```
This redefines the internal `id` function

illusional reviewed (changes requested):
Main points:

- We should chat about Python 3.10 (maybe at the stand-up / data meeting)
- I think the order of declarations within the `pipeline` file needs to be reworked
- It would be good to add the pre-commit hooks to this repo - saw a few black formatting and trailing whitespaces complaints from VSCode.

vladsaveliev commented:
Forgot to clarify that the code in this PR won't run in isolation from the rest of the codebase in the `initial` PR. This is just an attempt to split the PR for easier review. Otherwise it would be painful to merge. Let me know if it's okay.

vladsaveliev commented on cpg_pipes/pipeline/__init__.py:
```diff
@@ -0,0 +1,19 @@
+    DatasetStage, 
+    CohortStage,
+)
+from .targets import (
```
Forgot to clarify, this PR is not supposed to work in isolation, it's just meant to make the review process easier. These imports used in other parts of the codebase.

vladsaveliev commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+def stage(
+    _cls: Optional[Type['Stage']] = None, 
+    *,
+    analysis_type: str | None = None, 
```
That's right, it's meant to be for python 3.10 only. If/when we decide to make the code public for external users, we can add `__future__` imports to make it work with previous pythons.

vladsaveliev commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+        self, stage: StageDecorator
+    ) -> dict[str, dict[str, Path]]:
+        """
+        Get as a dictoinary of files/resources for a specific stage, indexed by target
```
Thanks! :)

vladsaveliev commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,960 @@
+
+
+def stage(
+    _cls: Optional[Type['Stage']] = None, 
```
Changed that

vladsaveliev commented:
> We should chat about Python 3.10 (maybe at the stand-up / data meeting)
Happy to chat, alternatively can add `__future__` imports
> I think the order of declarations within the pipeline file needs to be reworked
Changed the order as you suggested
> It would be good to add the pre-commit hooks to this repo - saw a few black formatting and trailing whitespaces complaints from VSCode.
Added the hooks. I disable black as most of its features are covered by PyCharm. Happy to run the entire code through black once all the PRs are in. Re whitespaces - I disabled that hook, hope it's okay.

illusional commented on .github/workflows/lint.yaml:
```diff
@@ -0,0 +1,29 @@
+    - name: Install packages
+      run: |
+        pip install .
+        pip install -r requirements-dev.txt
```
There is (currently) no requirements-dev.txt in this PR, so this is failing the linter.

lgruen commented:
> Forgot to clarify that the code in this PR won't run in isolation from the rest of the codebase in the `initial` PR. This is just an attempt to split the PR for easier review. Otherwise it would be painful to merge. Let me know if it's okay.

I think that's fine, as it'll really help to break down the individual reviews to manageable sizes without incurring a lot of overhead to make each PR work on its own.

illusional reviewed (approved):
LGTM!

lgruen commented on .flake8:
```diff
@@ -0,0 +1,25 @@
+# 10. F401 imported but unused (covered by PyCharm)
+extend-ignore: F541,E203,Q003,E501,W291,W293,E227,E125,F401
+
+# Allow listing module imports in __init__.py
```
I'm just curious, is that not allowed by default?

vladsaveliev replied:
It would report them as unused imports

vladsaveliev replied:
Alternatively I can add `__all__ = ['Pipeline', ...]` into `__init__.py` (see https://stackoverflow.com/questions/31079047/python-pep8-class-in-init-imported-but-not-used), but it's duplication

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+        @param batch: Hail Batch object
+        @param refs: reference data for bioinformatics
+        @param required_stages: list of stage classes that this stage requires
+        @param analysis_type: if defined, will query the SMDB Analysis entries 
```
> `will query the SMDB Analysis entries`

Is this now more abstract?

vladsaveliev replied:
Fixed!

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+        @param analysis_type: if defined, will query the SMDB Analysis entries 
+            of this type
+        @param skipped: means that the stage is skipped and self.queue_jobs()
+            won't run. The other stages if depend on it can aassume that that 
```
```suggestion
            won't run. The other stages if depend on it can assume that 
```

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+            of this type
+        @param skipped: means that the stage is skipped and self.queue_jobs()
+            won't run. The other stages if depend on it can aassume that that 
+            self.expected_result() returns existing files and target.ouptut_by_stage 
```
```suggestion
            self.expected_outputs() returns existing files and target.ouptut_by_stage 
```

vladsaveliev replied:
Renamed it everywhere

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+            won't run. The other stages if depend on it can aassume that that 
+            self.expected_result() returns existing files and target.ouptut_by_stage 
+            will be populated.
+        @param required: means that the self.expected_output() results are 
```
```suggestion
        @param required: means that the self.expected_outputs() results are 
```

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+        @param required: means that the self.expected_output() results are 
+            required for another active stage, even if the stage was skipped.
+        @param assume_outputs_exist: for skipped but required stages, 
+            the self.expected_result() output will still be checked for existence. 
```
```suggestion
            the self.expected_outputs() output will still be checked for existence. 
```

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+            the self.expected_result() output will still be checked for existence. 
+            This option makes the downstream stages assume that the output exist.
+        @param forced: run self.queue_jobs(), even if we can reuse the 
+            self.expected_output().
```
```suggestion
            self.expected_outputs().
```

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+        """
+
+    @abstractmethod
+    def expected_result(self, target: TargetT) -> ExpectedResultT:
```
Should this be called `expected_outputs`?

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+
+        self.status_reporter = None
+        input_provider = None
+        if (
+            input_provider_type == InputProviderType.SMDB 
+            or status_reporter_type == StatusReporterType.SMDB
+        ):
+            smdb = SMDB(self.cohort.analysis_dataset.name)
+            if status_reporter_type == StatusReporterType.SMDB:
+                self.status_reporter = SmdbStatusReporter(smdb)
+            if input_provider_type == InputProviderType.SMDB:
+                input_provider = SmdbInputsProvider(smdb)
```
I don't have a strong opinion about this, but you might prefer a design where the `Pipeline` class itself is completely agnostic to the underlying concrete implementations of the providers and only depends on the abstract interfaces. This way you avoid accidentally depending on concrete implementation details of any of the providers.

You'd then need to add a pipeline factory helper function (which can be completely isolated from the pipeline implementation itself), which instantiates those concrete providers and passes them to the pipeline.

vladsaveliev replied:
Good idea. Refactored it!

vladsaveliev replied:
Added a function `create_pipeline` (that can be run now with `cli_opts` instead of `Pipeline`), and simplified the `Pipeline` constructor to take already initialised `Cohort`, `Batch`, `StatusReporter`, and other objects. Let me know if it looks good.

lgruen commented on cpg_pipes/pipeline/pipeline.py:
```diff
@@ -0,0 +1,956 @@
+        # First round - checking which stages we require, even if they are skipped.
+        # If there are required stages that are not defined, we defined them
+        # into `additional_stages` as `skipped`.
+        # TODO: use TopologicalSorter
```
Is this a remnant that could now be addressed with Python 3.10 being supported?

vladsaveliev replied:
That's just a reminder to myself. Removed it!

lgruen reviewed (approved):
Looks great! Added a few comments that you might want to consider before merging.

lgruen commented on setup.py:
```diff
@@ -0,0 +1,50 @@
+        'fsspec',
+        'sample-metadata',
+        'analysis-runner',
+        'cloudpathlib[gs]',
```
`[gs]` or `[all]`?

vladsaveliev replied:
Fixed!

lgruen commented on setup.py:
```diff
@@ -0,0 +1,50 @@
+    classifiers=[
+        'Environment :: Console',
+        'Intended Audience :: Science/Research',
+        'License :: OSI Approved :: GNU General Public License v3 (GPLv3)',
```
This should be MIT.

Did we accidentally specify GPLv3 somewhere else in our repos?

vladsaveliev replied:
Fixed!
